host/sgx/sgxload.c

```diff
@@ -27,9 +27,68 @@
 #include <openenclave/internal/utils.h>
 #include "../memalign.h"
 #include "../signkey.h"
+#include "cpuid.h"
 #include "enclave.h"
 #include "sgxmeasure.h"
 
+#if defined(__linux__)
+// If cpuid.01h:ECX.XSAVE[bit 26] is 1, the processor supports the XSAVE/XRSTOR
+// processor extended states feature, the XSETBV/XGETBV instructions, and XCR0
+static bool is_xcr0_supported(void)
+{
+    uint32_t eax, ebx, ecx, edx;
+
+    eax = ebx = ecx = edx = 0;
+    oe_get_cpuid(1, 0, &eax, &ebx, &ecx, &edx);
+
+    // Bit 26 of ECX indicates support for XGETBV and XCR0
+    if (!(ecx & (1 << 26)))
```

Comment:
We should also check for bit 27 and optionally bit 28 here:



Bit | Name | Description
-- | -- | --
26 | XSAVE | A value of 1 indicates that the processor supports the XSAVE/XRSTOR processor extended states feature, the XSETBV/XGETBV instructions, and XCR0.
27 | OSXSAVE | A value of 1 indicates that the OS has set CR4.OSXSAVE[bit 18] to enable XSETBV/XGETBV instructions to access XCR0 and to support processor extended state management using XSAVE/XRSTOR.
28 | AVX | A value of 1 indicates the processor supports the AVX instruction extensions.

* Bit 27 is specifically the OS opt-in to XSAVE, which is rolled into the Windows `GetEnabledXStateFeatures()`
* Bit 28 doesn't technically need to be checked here, but can be a sanity check for your development if it is set but you're finding that `XGETBV` doesn't return bits 1 & 2 as set.

Reply:
I am checking for all these 3 bits in today's version of my code

Reply:
I am adding an extra sanity check that OS support for AVX matches the processor support. Thanks Simon.